File: smarthub/Home/src/services/ticketService.js
// src/services/ticketService.js

const API_BASE_URL = 'http://localhost:8081/api/tickets';

export const ticketService = {
  // Create a new ticket
  createTicket: async (ticketData) => {
    try {
      const response = await fetch(API_BASE_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(ticketData),
      });
      if (!response.ok) {
        throw new Error('Failed to create ticket');
      }
      return await response.json();
    } catch (error) {
      console.error('Error creating ticket:', error);
      throw error;
    }
  },

  // Get all tickets for a specific student
  getStudentTickets: async (studentId) => {
    try {
      const response = await fetch(`${API_BASE_URL}/student/${studentId}`);
      if (!response.ok) {
        throw new Error('Failed to fetch student tickets');
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching student tickets:', error);
      throw error;
    }
  },

  // Get all tickets (admin view)
  getAllTickets: async () => {
    try {
      const response = await fetch(API_BASE_URL);
      if (!response.ok) {
        throw new Error('Failed to fetch all tickets');
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching all tickets:', error);
      throw error;
    }
  },

  // Get a specific ticket
  getTicketById: async (ticketId) => {
    try {
      const response = await fetch(`${API_BASE_URL}/${ticketId}`);
      if (!response.ok) {
        throw new Error('Failed to fetch ticket');
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching ticket:', error);
      throw error;
    }
  },

  // Get tickets by status
  getTicketsByStatus: async (status) => {
    try {
      const response = await fetch(`${API_BASE_URL}/status/${status}`);
      if (!response.ok) {
        throw new Error('Failed to fetch tickets by status');
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching tickets by status:', error);
      throw error;
    }
  },

  // Update ticket status
  updateTicketStatus: async (ticketId, status) => {
    try {
      const response = await fetch(`${API_BASE_URL}/${ticketId}/status?status=${status}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
      });
      if (!response.ok) {
        throw new Error('Failed to update ticket status');
      }
      return await response.json();
    } catch (error) {
      console.error('Error updating ticket status:', error);
      throw error;
    }
  },

  // Add admin comments
  addAdminComments: async (ticketId, comments) => {
    try {
      const response = await fetch(`${API_BASE_URL}/${ticketId}/comments?comments=${encodeURIComponent(comments)}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
      });
      if (!response.ok) {
        throw new Error('Failed to add admin comments');
      }
      return await response.json();
    } catch (error) {
      console.error('Error adding admin comments:', error);
      throw error;
    }
  },

  // Delete ticket
  deleteTicket: async (ticketId) => {
    try {
      const response = await fetch(`${API_BASE_URL}/${ticketId}`, {
        method: 'DELETE',
      });
      if (!response.ok) {
        throw new Error('Failed to delete ticket');
      }
      return true;
    } catch (error) {
      console.error('Error deleting ticket:', error);
      throw error;
    }
  },
};

File: smarthub/Backend/src/main/java/com/paf/backend/service/TicketService.java
package com.paf.backend.service;

import com.paf.backend.dto.TicketDTO;
import com.paf.backend.entity.Ticket;
import com.paf.backend.repository.TicketRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class TicketService {
    
    @Autowired
    private TicketRepository ticketRepository;
    
    // Create a new ticket
    public TicketDTO createTicket(TicketDTO ticketDTO) {
        Ticket ticket = new Ticket();
        ticket.setStudentId(ticketDTO.getStudentId());
        ticket.setTitle(ticketDTO.getTitle());
        ticket.setDescription(ticketDTO.getDescription());
        ticket.setCategory(Ticket.Category.valueOf(ticketDTO.getCategory()));
        ticket.setPriority(Ticket.Priority.valueOf(ticketDTO.getPriority()));
        ticket.setStatus(Ticket.Status.OPEN);
        ticket.setCreatedAt(LocalDateTime.now());
        
        Ticket savedTicket = ticketRepository.save(ticket);
        return new TicketDTO(savedTicket);
    }
    
    // Get all tickets for a student
    public List<TicketDTO> getStudentTickets(String studentId) {
        List<Ticket> tickets = ticketRepository.findByStudentId(studentId);
        return tickets.stream().map(TicketDTO::new).collect(Collectors.toList());
    }
    
    // Get a specific ticket
    public Optional<TicketDTO> getTicketById(String id) {
        Optional<Ticket> ticket = ticketRepository.findById(id);
        return ticket.map(TicketDTO::new);
    }
    
    // Get all tickets by status
    public List<TicketDTO> getTicketsByStatus(String status) {
        List<Ticket> tickets = ticketRepository.findByStatus(Ticket.Status.valueOf(status));
        return tickets.stream().map(TicketDTO::new).collect(Collectors.toList());
    }
    
    // Get all tickets
    public List<TicketDTO> getAllTickets() {
        List<Ticket> tickets = ticketRepository.findAll();
        return tickets.stream().map(TicketDTO::new).collect(Collectors.toList());
    }
    
    // Update ticket status
    public Optional<TicketDTO> updateTicketStatus(String id, String status) {
        Optional<Ticket> ticketOpt = ticketRepository.findById(id);
        if (ticketOpt.isPresent()) {
            Ticket ticket = ticketOpt.get();
            ticket.setStatus(Ticket.Status.valueOf(status));
            ticket.setUpdatedAt(LocalDateTime.now());
            Ticket updatedTicket = ticketRepository.save(ticket);
            return Optional.of(new TicketDTO(updatedTicket));
        }
        return Optional.empty();
    }
    
    // Add admin comments to ticket
    public Optional<TicketDTO> addAdminComments(String id, String comments) {
        Optional<Ticket> ticketOpt = ticketRepository.findById(id);
        if (ticketOpt.isPresent()) {
            Ticket ticket = ticketOpt.get();
            ticket.setAdminComments(comments);
            ticket.setUpdatedAt(LocalDateTime.now());
            Ticket updatedTicket = ticketRepository.save(ticket);
            return Optional.of(new TicketDTO(updatedTicket));
        }
        return Optional.empty();
    }
    
    // Delete ticket
    public boolean deleteTicket(String id) {
        if (ticketRepository.existsById(id)) {
            ticketRepository.deleteById(id);
            return true;
        }
        return false;
    }
}
